frame/mixnet/src/lib.rs

```diff
@@ -0,0 +1,675 @@
+	for BoundedMixnode<BoundedVec<BoundedVec<u8, MaxExternalAddressSize>, MaxExternalAddresses>>
+{
+	fn from(mixnode: Mixnode) -> Self {
+		Self {
+			kx_public: mixnode.kx_public,
+			peer_id: mixnode.peer_id,
+			external_addresses: mixnode
+				.external_addresses
+				.into_iter()
+				.flat_map(|addr| match addr.try_into() {
+					Ok(addr) => Some(addr),
+					Err(addr) => {
+						log::debug!(target: LOG_TARGET,
+							"Mixnode external address {addr:x?} too long; discarding");
+						None
+					},
+				})
+				.take(MaxExternalAddresses::get() as usize)
+				.collect::<Vec<_>>()
+				.try_into()
+				.expect("Excess external addresses discarded with take()"),
+		}
+	}
+}
+
+/// [`BoundedMixnode`] type for the given configuration.
+pub type BoundedMixnodeFor<T> = BoundedMixnode<
+	BoundedVec<
+		BoundedVec<u8, <T as Config>::MaxExternalAddressSize>,
+		<T as Config>::MaxExternalAddressesPerMixnode,
+	>,
+>;
+
+////////////////////////////////////////////////////////////////////////////////
+// Registration type
+////////////////////////////////////////////////////////////////////////////////
+
+/// A mixnode registration. A registration transaction is formed from one of these plus an
+/// [`AuthoritySignature`].
+#[derive(Clone, Decode, Encode, PartialEq, TypeInfo, RuntimeDebug)]
+pub struct Registration<BlockNumber, BoundedMixnode> {
+	/// Block number at the time of creation. When a registration transaction fails to make it on
+	/// to the chain for whatever reason, we send out another one. We want this one to have a
+	/// different hash in case the earlier transaction got banned somewhere; including the block
+	/// number is a simple way of achieving this.
+	pub block_number: BlockNumber,
+	/// The session during which this registration should be processed. Note that on success the
+	/// mixnode is registered for the _following_ session.
+	pub session_index: SessionIndex,
+	/// The index in the next session's authority list of the authority registering the mixnode.
+	pub authority_index: AuthorityIndex,
+	/// Mixnode information to register for the following session.
+	pub mixnode: BoundedMixnode,
+}
+
+/// [`Registration`] type for the given configuration.
+pub type RegistrationFor<T> = Registration<BlockNumberFor<T>, BoundedMixnodeFor<T>>;
+
+////////////////////////////////////////////////////////////////////////////////
+// Misc helper funcs
+////////////////////////////////////////////////////////////////////////////////
+
+fn check_removed_all(res: MultiRemovalResults) {
+	debug_assert!(res.maybe_cursor.is_none());
+}
+
+fn twox<BlockNumber: UniqueSaturatedInto<u64>>(
+	block_number: BlockNumber,
+	kx_public: &KxPublic,
+) -> u64 {
+	let block_number: u64 = block_number.unique_saturated_into();
+	let mut data = [0; 8 + KX_PUBLIC_SIZE];
+	*array_mut_ref![data, 0, 8] = block_number.to_le_bytes();
+	*array_mut_ref![data, 8, KX_PUBLIC_SIZE] = *kx_public;
```

Comment:
```suggestion
	data[..8].copy_from_slice(block_number.to_le_bytes());
	data[8..].copy_from_slice(&*kx_public);
```

No need to import this extra crate.

Reply:
Done.